include/aws/s3/private/s3_request.h

```diff
@@ -23,6 +24,69 @@ enum aws_s3_request_flags {
     AWS_S3_REQUEST_FLAG_ALWAYS_SEND = 0x00000004,
 };
 
+/**
+ * Information sent in the meta_request telemetry callback.
+ */
+struct aws_s3_request_metrics {
+    struct aws_allocator *allocator;
+
+    struct {
+        /* The time stamp when the request started by S3 client, which is prepared time by the client. Timestamps
+         * are from `aws_high_res_clock_get_ticks`. This will always be available. */
+        uint64_t start_timestamp_ns;
+        /* The time stamp when the request finished by S3 client succeed or failed or to be retried. Timestamps
+         * are from `aws_high_res_clock_get_ticks`. This will always be available. */
+        uint64_t end_timestamp_ns;
+        /* The time duration for the request from start to finish. end_timestamp_ns - start_timestamp_ns. This will
+         * always be available. */
+        uint64_t total_duration_ns;
+
+        /* The time stamp when the request started to be encoded. 0 means data not available. Timestamp
+         * are from `aws_high_res_clock_get_ticks` */
+        uint64_t send_start_timestamp_ns;
+        /* The time stamp when the request finished to be encoded. 0 means data not available.
+         * Timestamp are from `aws_high_res_clock_get_ticks` */
+        uint64_t send_end_timestamp_ns;
+        /* The time duration for the request from start encoding to finish encoding (send_end_timestamp_ns -
+         * send_start_timestamp_ns). When send_end_timestamp_ns is 0, means data not available. */
+        uint64_t sending_duration_ns;
+
+        /* The time stamp when the response started to be received from the network channel. 0 means data not
+         * available. Timestamp are from `aws_high_res_clock_get_ticks` */
+        uint64_t receive_start_timestamp_ns;
+        /* The time stamp when the response finished to be received from the network channel. 0 means data not
+         * available. Timestamp are from `aws_high_res_clock_get_ticks` */
+        uint64_t receive_end_timestamp_ns;
+        /* The time duration for the request from start receiving to finish receiving. receive_end_timestamp_ns -
+         * receive_start_timestamp_ns. When receive_end_timestamp_ns is 0, means data not available. */
+        uint64_t receiving_duration_ns;
+    } time_metrics;
+
+    struct {
+        /* Response status code for the request */
+        int response_status;
+        /* HTTP Headers of the response received. */
+        struct aws_http_headers *response_headers;
+        /* The request line: method, path and query of the request. */
+        struct aws_byte_buf request_path_query;
```

Comment:
should we add the whole url? with virtual host addressing, path will not have bucket

Reply:
I am not sure, the virtual host address seems similar to the ip address, and bucket is kind provided by user?

Reply:
virtual host address is just bucket being subdomain. bucket is provided by user, but then user needs to do mapping to figure out complete endpoint this request is addressing. adding whole url would just make it simpler for user which request metric is associated with

Reply:
So, should it be the host header of the request? I can add that